File: backend/src/services/deck.service.ts
import { GameState } from '../models/game-state';
import { Card } from '../models/card';
import { Companion } from '../models/companion';

/**
 * Pure deck-management service.
 *
 * All methods are stateless: they accept a GameState and return a new one
 * (or return plain data structures). No file I/O; no side effects.
 */
export class DeckService {

  // ---------------------------------------------------------------------------
  // Shuffle
  // ---------------------------------------------------------------------------

  /** Fisher-Yates shuffle. Returns a new array; never mutates the input. */
  shuffle<T>(arr: T[]): T[] {
    const a = arr.slice();
    for (let i = a.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [a[i], a[j]] = [a[j], a[i]];
    }
    return a;
  }

  // ---------------------------------------------------------------------------
  // Draw
  // ---------------------------------------------------------------------------

  /**
   * Draw `count` cards from the deck into the hand.
   *
   * If the deck is exhausted mid-draw the discard pile is shuffled back in
   * automatically before continuing.  If both are empty the draw stops early.
   *
   * Returns a new GameState with updated hand / deck / discard.
   */
  drawCards(state: GameState, count: number): GameState {
    let deck    = state.player.deck.cardIds.slice();
    let hand    = state.player.hand.slice();
    let discard = (state.player.discard ?? []).slice();

    for (let i = 0; i < count; i++) {
      if (deck.length === 0) {
        if (discard.length === 0) {
          // Nothing left anywhere — stop drawing.
          break;
        }
        // Shuffle discard back into deck.
        deck    = this.shuffle(discard);
        discard = [];
      }
      hand.push(deck.shift()!);
    }

    return {
      ...state,
      player: {
        ...state.player,
        hand,
        deck:    { ...state.player.deck, cardIds: deck },
        discard,
      },
    };
  }

  // ---------------------------------------------------------------------------
  // Starting deck builder
  // ---------------------------------------------------------------------------

  /**
   * Build the master card catalogue and starting deck for a new run.
   *
   * Combines `baseCards` with two companion-specific starter cards per companion.
   * Returns both the full `cards` array (for `GameState.cards`) and a `deck`
   * whose `cardIds` list every card id in draw order (unshuffled — shuffle on
   * battle start if desired).
   */
  buildStartingDeck(
    baseCards: Card[],
    companions: Companion[]
  ): { deck: { cardIds: string[] }; cards: Card[] } {
    const cards: Card[] = baseCards.slice();

    companions.forEach((companion, idx) => {
      // Attack-type starter card
      cards.push({
        id:              `comp-${companion.id}-${idx}-a`,
        name:            `${companion.name} Strike`,
        cost:            1,
        type:            companion.type,
        description:     'Companion basic attack',
        effectId:        'fx-comp-strike-normal',
        enhancedEffectId:'fx-comp-strike-enhanced',
        effect:          { description: 'Deal 4 damage to one enemy.' },
        enhancedEffect:  { description: 'Deal 6 damage to one enemy.' },
      });

      // Defence starter card
      cards.push({
        id:              `comp-${companion.id}-${idx}-b`,
        name:            `${companion.name} Guard`,
        cost:            1,
        type:            'defense',
        description:     'Companion basic defence',
        effectId:        'fx-comp-guard-normal',
        enhancedEffectId:'fx-comp-guard-enhanced',
        effect:          { description: 'Gain 3 shield.' },
        enhancedEffect:  { description: 'Gain 5 shield.' },
      });
    });

    return {
      deck:  { cardIds: cards.map(c => c.id) },
      cards,
    };
  }
}

File: backend-csharp/TheGameIsOn.API/Services/DeckService.cs
using TheGameIsOn.API.Models;

namespace TheGameIsOn.API.Services;

public interface IDeckService
{
    List<T> Shuffle<T>(List<T> list);
    GameState DrawCards(GameState state, int count);
    (Deck Deck, List<Card> Cards) BuildStartingDeck(List<Card> baseCards, List<Companion> companions);
}

/// <summary>
/// Pure deck-management service.
/// All methods are stateless: they accept a GameState and return a new one
/// (or return plain data structures). No file I/O; no side effects.
/// </summary>
public class DeckService : IDeckService
{
    /// <summary>Fisher-Yates shuffle. Returns a new list; never mutates the input.</summary>
    public List<T> Shuffle<T>(List<T> list)
    {
        var a = new List<T>(list);
        for (var i = a.Count - 1; i > 0; i--)
        {
            var j = Random.Shared.Next(i + 1);
            (a[i], a[j]) = (a[j], a[i]);
        }
        return a;
    }

    /// <summary>
    /// Draw <paramref name="count"/> cards from the deck into the hand.
    /// If the deck is exhausted mid-draw the discard pile is shuffled back in
    /// automatically before continuing. If both are empty the draw stops early.
    /// Returns a new GameState with updated hand / deck / discard.
    /// </summary>
    public GameState DrawCards(GameState state, int count)
    {
        var deck    = new List<string>(state.Player.Deck.CardIds);
        var hand    = new List<string>(state.Player.Hand);
        var discard = new List<string>(state.Player.Discard);

        for (var i = 0; i < count; i++)
        {
            if (deck.Count == 0)
            {
                if (discard.Count == 0) break;
                deck = Shuffle(discard);
                discard = new List<string>();
            }

            hand.Add(deck[0]);
            deck.RemoveAt(0);
        }

        return new GameState
        {
            Player = state.Player with
            {
                Hand = hand,
                Deck = state.Player.Deck with { CardIds = deck },
                Discard = discard
            },
            Graph = state.Graph,
            Cards = state.Cards,
            Companions = state.Companions,
            History = state.History,
            Battle = state.Battle,
            PendingAbilityChoices = state.PendingAbilityChoices
        };
    }

    /// <summary>
    /// Build the master card catalogue and starting deck for a new run.
    /// Combines <paramref name="baseCards"/> with two companion-specific
    /// starter cards per companion.
    /// </summary>
    public (Deck Deck, List<Card> Cards) BuildStartingDeck(List<Card> baseCards, List<Companion> companions)
    {
        var cards = new List<Card>(baseCards);

        for (var idx = 0; idx < companions.Count; idx++)
        {
            var companion = companions[idx];
            var companionElement = companion.Element ?? CardElement.Neutral;

            // Attack-type starter card — inherits the companion's element
            cards.Add(new Card
            {
                Id               = $"comp-{companion.Id}-{idx}-a",
                Name             = $"{companion.Name} Strike",
                Cost             = 1,
                Type             = companion.Type,
                Element          = companionElement,
                Description      = "Companion basic attack",
                EffectId         = "fx-comp-strike-normal",
                EnhancedEffectId = "fx-comp-strike-enhanced",
                Effect           = new CardEffectRef("Deal 4 damage to one enemy."),
                EnhancedEffect   = new CardEffectRef("Deal 6 damage to one enemy.")
            });

            // Defence starter card — inherits the companion's element
            cards.Add(new Card
            {
                Id               = $"comp-{companion.Id}-{idx}-b",
                Name             = $"{companion.Name} Guard",
                Cost             = 1,
                Type             = CardType.Defense,
                Element          = companionElement,
                Description      = "Companion basic defence",
                EffectId         = "fx-comp-guard-normal",
                EnhancedEffectId = "fx-comp-guard-enhanced",
                Effect           = new CardEffectRef("Gain 3 shield."),
                EnhancedEffect   = new CardEffectRef("Gain 5 shield.")
            });
        }

        var deck = new Deck { CardIds = cards.Select(c => c.Id).ToList() };
        return (deck, cards);
    }
}
